### ISSUE TITLE
共享的embedding作为下游Layer的输入，动态图模式下如何训练embedding 的权重

### ISSUE THREAD

**Issue by ppy-hellowd** [PaddlePaddle/Paddle#37650]
构建了一个输入embedding Layer，作为下游两个 Layer的输入，得到两个loss，再用两个 optimizer 优化。Optimizer 不知道该怎么设置parameter_list，一个 optimizer能同时优化上下游两个layer吗？

    def main(params):
        with D.guard():
            # data batch reader
            data_reader = MyDataReader(FLIST)
            # layers
            c_embed = StateEmbedding(params, name_scope="c_state_emb")
            n_embed = StateEmbedding(params, name_scope="n_state_emb")
            actor = Actor(params)
            actor_n = Actor(params, trainable=False)
            critic = CriticNet(params)
            critic_n = CriticNet(params, trainable=False)
            # train op
            actor_op = OP.AdagradOptimizer(learning_rate=params['actor_learning_rate'], parameter_list=actor.parameters())
            critic_op = OP.AdagradOptimizer(learning_rate=params['critic_learning_rate'], parameter_list=critic.parameters())
    
            for i in range(params["max_steps"]):
                S1, A1, R1, S2 = data_reader.next_batch()
                s1, a1, r1, s2 = Get_variable(S1, A1, R1, S2)
    
                c_state_embed = c_embed(s1[0], s1[1], s1[2], s1[3], s1[4], s1[5], s1[6], s1[7])
                n_state_embed = n_embed(s2[0], s2[1], s2[2], s2[3], s2[4], s2[5], s2[6], s2[7])
                action_pred   = actor(c_state_embed) # predict action
                action_next   = actor_n(n_state_embed) # next action
                
                q_pred_for_actor  = critic(c_state_embed, action_pred) # predict Q value for actor to update 
                q_pred_for_critic = critic(c_state_embed, a1) # predict Q value for critic to update
                q_next  = critic_n(n_state_embed, action_next)
                q_label = L.sums(input=[params['gamma'] * q_next, r1])
    
                critic_loss_vec = L.square_error_cost(input=q_pred_for_critic, label=q_label)
                actor_loss_vec  = (-1) * q_pred_for_actor
                critic_loss     = L.reduce_mean(input=critic_loss_vec)
                actor_loss      = L.reduce_mean(input=actor_loss_vec)
                actor_loss.backward()
                critic_loss.backward()
                # minimize
                actor_op.minimize(actor_loss)
                critic_op.minimize(critic_loss)
                # clear gradients
                c_embed.clear_gradients() # train embedding weights
                actor.clear_gradients()
                critic.clear_gradients()

**Comment by phlrain** [comment]
请问下，是期望两个optimize都更新同一个参数么

**Comment by ppy-hellowd** [comment]
一个共享的 embeddingLayer，和 后面拼接的两个独立的 layer。我现在定义的 optimizer 是只更新了那两个独立的 layer，我怎么根据分别的 loss，更新 embedding layer

**Comment by JiabinYang** [comment]
> 一个共享的 embeddingLayer，和 后面拼接的两个独立的 layer。我现在定义的 optimizer 是只更新了那两个独立的 layer，我怎么根据分别的 loss，更新 embedding layer

这两个optimizer是希望怎样的方式更新同一个参数呢，是希望optimizerA更新一下，然后optimizerB更新一下 还是根据什么条件来区分只用optimizerA更新或者只用opitmizerB更新呢

**Comment by ppy-hellowd** [comment]
我希望可以自定义更新方式，在需要的时候更新，不需要的时候冻结

**Comment by JiabinYang** [comment]
> 我希望可以自定义更新方式，在需要的时候更新，不需要的时候冻结

那你可以根据实际的判断情况选择性的把embedding的参数放进optimizer需要的参数列表里面

**Comment by ppy-hellowd** [comment]
尝试过：
actor_op = OP.AdagradOptimizer(learning_rate=params['actor_learning_rate'], parameter_list=actor.parameters())
设置 parameter_list=embedding.parameters()+actor.parameters()，这样写结果报错了

**Comment by JiabinYang** [comment]
> parameter_list

具体报什么错误可以贴一下我们看看
